cpp/src/io/functions.cpp

```diff
@@ -191,6 +193,88 @@ void write_csv(csv_writer_options const& options, rmm::mr::device_memory_resourc
 
 namespace detail_orc = cudf::io::detail::orc;
 
+// Freeform API wraps the detail reader class API
+std::vector<std::vector<std::string>> read_orc_statistics(source_info const& src_info)
+{
+  // Get source to read statistics from
+  std::unique_ptr<datasource> source;
+  if (src_info.type == io_type::FILEPATH) {
+    CUDF_EXPECTS(src_info.filepaths.size() == 1, "Only a single source is currently supported.");
+    source = cudf::io::datasource::create(src_info.filepaths[0]);
+  } else if (src_info.type == io_type::HOST_BUFFER) {
+    CUDF_EXPECTS(src_info.buffers.size() == 1, "Only a single source is currently supported.");
+    source = cudf::io::datasource::create(src_info.buffers[0]);
+  } else if (src_info.type == io_type::USER_IMPLEMENTED) {
+    CUDF_EXPECTS(src_info.user_sources.size() == 1, "Only a single source is currently supported.");
+    source = cudf::io::datasource::create(src_info.user_sources[0]);
+  } else {
+    CUDF_FAIL("Unsupported source type");
+  }
```

Comment:
This could be done as IILE:
```
std::unique_ptr<datasource> source = [&] {
  if (src_info.type == io_type::FILEPATH) {
    CUDF_EXPECTS(src_info.filepaths.size() == 1, "Only a single source is currently supported.");
    return cudf::io::datasource::create(src_info.filepaths[0]);
  }
  if (src_info.type == io_type::HOST_BUFFER) {
    CUDF_EXPECTS(src_info.buffers.size() == 1, "Only a single source is currently supported.");
    return cudf::io::datasource::create(src_info.buffers[0]);
  } 
  if (src_info.type == io_type::USER_IMPLEMENTED) {
    CUDF_EXPECTS(src_info.user_sources.size() == 1, "Only a single source is currently supported.");
    return cudf::io::datasource::create(src_info.user_sources[0]);
  }
  CUDF_FAIL("Unsupported source type");
 }();
```